// vdna/variable_data.js
module.exports = {
  facebook: [
    {
      cnet:        { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                     related: 'technology,helth,science' },
      "jazz dock": { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                     related: 'music,jazz,concerts' }
    },
    {
      "new scientist": { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                         related: 'science,literature' },
      "divadlo archa": { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                         related: 'music,theater,concerts' }
    },
    {
      hiking:    { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                   related: 'health,fitness' },
      recycling: { source: 'facebook', clicks: 0, added: Date.now(), selected: true,
                   related: 'technology' }
    }
  ],
  newLikes: {},
  totalFacebookSync: 0,
  pinterest: [
    {
      mammals: { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                 related: 'science' },
      spain:   { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                 related: 'literature,film,music' }
    },
    {
      prague:  { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                 related: 'literature,film,music' },
      ferrets: { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                 related: 'humor,science' }
    },
    {
      "avant garde":   { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                         related: 'music,drama,theater,film' },
      synchronicities: { source: 'pintrest', clicks: 0, added: Date.now(), selected: true,
                         related: 'spirituality,literature,film' }
    }
  ],
  totalPinterestSync: 0,

  facebookImportReset: function() {
    this.facebook = [];
    this.totalFacebookSync = 0;
  },

  importNewLike: function(like) {
    this.newLikes[like.toLowerCase()] = { source: 'facebook', clicks: 0, added: Date.now(), selected: true, related: '' };
  },

  pushNewLikes: function() {
    if(Object.keys(this.newLikes).length > 0) {
      this.facebook.unshift(this.newLikes);
      this.newLikes = {};
    }
  }
};
